chia/full_node/full_node_api.py

```diff
@@ -1103,15 +1107,37 @@ async def request_block_header(self, request: wallet_protocol.RequestBlockHeader
             msg = make_msg(ProtocolMessageTypes.reject_header_request, RejectHeaderRequest(request.height))
             return msg
         block: Optional[FullBlock] = await self.full_node.block_store.get_full_block(header_hash)
-        if block is not None:
-            tx_removals, tx_additions, _ = await self.full_node.blockchain.get_tx_removals_and_additions(block)
-            header_block = get_block_header(block, tx_additions, tx_removals)
-            msg = make_msg(
-                ProtocolMessageTypes.respond_block_header,
-                wallet_protocol.RespondBlockHeader(header_block),
+        if block is None:
+            return None
+
+        tx_removals: List[bytes32] = []
+        tx_additions: List[Coin] = []
+
+        if block.transactions_generator:
+
+            block_generator: Optional[BlockGenerator] = await self.full_node.blockchain.get_block_generator(block)
+            assert block_generator is not None
+
+            npc_result = await asyncio.get_running_loop().run_in_executor(
+                self.executor,
+                lambda x: get_name_puzzle_conditions(
+                    x,
+                    self.full_node.constants.MAX_BLOCK_COST_CLVM,
+                    cost_per_byte=self.full_node.constants.COST_PER_BYTE,
+                    mempool_mode=False,
+                ),
+                block_generator,
```

Comment:
I prefer to avoid closures.  Regardless, the lambda was an interesting mix of passing one positional arg through the executor run call which can take several and others through closures from the lambda.  Could also pass all parameters by name with `functools.partial()`.

```suggestion
                functools.partial(
                    get_name_puzzle_conditions,
                    block_generator,
                    self.full_node.constants.MAX_BLOCK_COST_CLVM,
                    cost_per_byte=self.full_node.constants.COST_PER_BYTE,
                    mempool_mode=False,
                ),
```

Reply:
are you sure this works? if I'm reading the documentation correctly, you bind arguments from the beginning of the parameter list, not the end.

Reply:
I don't think I understand the concern.

https://github.com/Chia-Network/chia-blockchain/blob/d66abf04436db9129413c5ab2f5f88a9ce52418c/chia/full_node/mempool_check_conditions.py#L33-L35

So a 'regular' call would be like this, right?

```python
get_name_puzzle_conditions(
    block_generator,
    self.full_node.constants.MAX_BLOCK_COST_CLVM,
    cost_per_byte=self.full_node.constants.COST_PER_BYTE,
    mempool_mode=False,
)
```

So the same order would apply with `functools.partial()`.  Though just using names could alleviate order concerns everywhere.

```python-console
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import functools
>>> def f(a, b):
...     print(a, b)
... 
>>> p = functools.partial(f, 1, 2)
>>> p()
1 2
```

Reply:
the concern is demonstrated by this example:
```
Python 3.9.14 (main, Sep  6 2022, 23:16:16) 
[Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import functools
>>> def f(a, b):
...     print(a, b)
... 
>>> p = functools.partial(f, 1)
>>> p(2)
1 2
```

In the code above, we need to bind the *last* 3 parameters of `get_name_puzzle_conditions()`, not the *first* 3.

Reply:
I suppose we could bind all parameters. The risk then is to make sure `block_generator` is transferred efficiently into the thread. But since this is just a thread, and not a process, presumably it's just transferred by reference, so it wouldn't be an issue.

Reply:
I'm curious; what's wrong with closures?

Reply:
Closures make reference tracking and teardown more complicated and error prone I believe and also are more likely to confuse developers when used in loops (not here, but in general) due to their late binding behavior.  If you are interested in more details about either case maybe a chat would be good.

Yes, `functools.partial()` does bind as you expected and as you see.  For that exact topic alone, fix it by passing by name even for the parameters that don't require it.  The thing is that there was no need to be passing some positional items via the executor and some by closure and the call in the lambda.  Since some are keyword only, you can't pass them via the executor since it only handles positional arguments.  So they ought have just gone all into the lambda.  But then there are even more closures, hence the partial suggestion.

While not quite the core of the topic here, the 'here is a callable and arguments will be provided later' topic is generally kind of hard.  It makes for a complicated API to figure out how to have a callable like `.run_in_executor()` take it's own positional and keyword args while also taking the positional and keyword args for the callable it is passed.  We could just handle the callable's parameters via lambdas or partials and take _none_ via the executor call.  Could have the executor take it's own and return a callable that takes the other params like `.run_in_executor(the_function, an_executor_argument)(the_function_argument_1, the_function_argument_2)`.  Could do as it does and just discourage people from using the clarity of keyword arguments thus maximizing bug risk.  For a particularly extended discussion you could check out https://github.com/python-trio/trio/issues/470.

Reply:
Your concern about order is a bit more critical when it comes to a map-like function where the callable may be called multiple times with different parameters and you can't include _all_ parameters in the partial.  In that case, yes, you have to deal with the order when the map-like can't handle keyword arguments.  I would default to using keyword arguments to create the partial.